### DISCUSSION TITLE
How to ignore properties during EDM build

### DISCUSSION THREAD

**Discussion by RuiHou105** [dotnet/aspnet-api-versioning#1020] [Q&A]
Hi team, 

I know that if we want to ignore properties during EDM model building, we can use the format: for example, `auditEvents.Ignore(t => t.activity)`. In this way the property `activity` is ignored. 

But right now, I am working on providing a common method for service teams to ignore their unused properties. So, in my method, the usage can be: `generatedObject.Ignore( t => t.propertyName)`. My question is: since this is a common method, actually I don't know every class's property name, so how can I write the expression in `Ignore `method? Or is there any other way to ignore properties?

Thanks,
Rui

**Comment by commonsensesoftware** [comment]
Is there a reason why you could not just have model authors apply `[JsonIgnore]` (OData 8+) or `[IgnoreDataMember]` to the properties they want to have excluded? It seems that would solve the problem without having to write any special code on your own.

**Reply by RuiHou105** [reply]
Hi @commonsensesoftware  thanks for your reply! Here My scenario is: firstly, use the API versioning tool to specify which graph version current controller supports. If current supported version is `beta`, I am going to ignore **nothing**; if current supported version is `V1.0`, I am going to ignore some properties in the startup. **So here I can't directly add [`JsonIgnore`] attribute for properties, since beta version won't ignore any properties.**

**Reply by commonsensesoftware** [reply]
There are multiple ways to achieve your result, but you'll likely need to create/expose a facility to teams that you can consume. An attribute or interface are good candidates, but that's not the only way.

Let's say you added the attribute:

```c#
[AttributeUsage(AttributeTargets.Property)]
public sealed class InPreviewAttribute : Attribute
{
}
```

Then clients might apply it as:

```c#
public class Person
{
    public int Id { get; set; }
    
    public string FirstName { get; set; }

    public string LastName { get; set; }

    [InPreview]
    public string Email { get; set; }
}
```

I presume you must be using Reflection to enumerate the properties. A rough skeleton _might_ look like:

```c#
var builder = new ODataModelConventionBuilder();
var type = typeof(Person);
var entity = builder.AddEntitySet("People", type).EntityType;
var preview = !string.IsNullOrEmpty(apiVersion.Status);

if (preview)
{
    foreach (var property in type.GetRuntimeProperties())
    {
        if (property.GetCustomAttribute<InPreviewAttribute>() is not null)
        {
            entity.RemoveProperty(property);
        }
    }
}
```

If you know the clients at build-time, then you can use a source generator instead of Reflection. You could also have the clients implement a particular interface or provide some other method to enable them to filter properties. Avoiding Reflection will improve your startup time.
